torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -570,23 +576,42 @@ def moe_forward(self, x, topk_ids, topk_weight):
             )
 
         # DP to EP token shuffle. This part needs gradient.
-        on_device_all_to_all_v(
-            MoE.token_gather_buf,
-            MoE.output_splits,
-            sorted_tokens,
-            MoE.input_splits,
-            self.ep_group,
-        )
-
-        with torch.no_grad():
-            # Output splits sanity check
-            # expected_splits = tokens_per_expert_group.view(self.ep_size, -1).sum(dim=1)
-            # torch.testing.assert_close(MoE.output_splits, expected_splits)
-            # Received tokens from all other ranks. TODO: use mask instead
-            received = MoE.output_splits.sum()
-
-        # TODO: don't use `received`
-        gathered_tokens = MoE.token_gather_buf[:received]
+        if self.shuffle_method == "symm_mem":
+            # Move input to the `token_send_buf` symm mem
+            MoE.token_send_buf[: idxs.shape[0]].copy_(sorted_tokens)
+            # Note: `out=` avoids copy, but it is not differentiable
+            # torch.index_select(x, 0, idxs // topk_ids.shape[1], out=MoE.token_send_buf[: idxs.shape[0]])
+            with torch.no_grad():
+                torch.sum(
+                    tokens_per_expert.view(self.ep_size, -1),
+                    dim=1,
+                    out=MoE.input_splits,
+                )
+            on_device_all_to_all_v(
+                MoE.token_gather_buf,
+                MoE.output_splits,
+                MoE.token_send_buf,
+                MoE.input_splits,
+                self.ep_group,
+            )
+            with torch.no_grad():
+                # Received tokens from all other ranks. TODO: use mask instead
+                received = MoE.output_splits.sum()
+            # TODO: don't use `received`
+            gathered_tokens = MoE.token_gather_buf[:received]
+        else:  # "torch_all_to_all"
+            # Prepare input ans output splits
```

Comment:
Typo: and

also, the point of this else branch is just to offer  EP>8 support for now, right?  and perhaps also benchmark between the two on EP=8